### ISSUE TITLE
Login with consent

### ISSUE THREAD

**Issue by David-hod** [panva/node-oidc-provider#442]
I think it can really help for new developers to create full examples for interaction since it is not really understood from the docs.
when I say example i mean : 
1. login + consent  using post/submit
2. login + consent using ajax.

I am trying to debug the code of  `interactionResult()`  and `interactionDetails()` in order to understand how I should write both flows. it also would be good to implement error handling on those examples. i.e. user does not agree to the consent or user with wrong credentials. 

I know it may take time to write those example, but it would be much more friendly to new Developers that really struggle with it.

**Comment by panva** [comment]
I'm not opposed to more detailed examples, but,

- XHR i won't entertain
- Failed submissions should not reach the interaction result, just re-render your views like you would in a regular standalone application. The only thing the interaction helpers give you is context (details) and means to complete the flow when your own conditions are satisfied.

I kinda feel like i'm asked to provide examples of how to do get/post forms in express and koa, that is a prerequisite when working with web frameworks.

**Comment by David-hod** [comment]
> I kinda feel like i'm asked to provide examples of how to do get/post forms in express and koa, that is a prerequisite when working with web frameworks.
it's not about using get/post its about using it's about how to use `interactionResult(),interactionDetails() ` inside those handlers.

**Comment by panva** [comment]
ask those questions then with as many details as possible so i can help, i can hardly provide an example for everything.

> I don't have any clue about what is really happening there. what is the life cycle of the cookie/session?

This [doc section](https://github.com/panva/node-oidc-provider/blob/master/docs/configuration.md#interaction) sheds some light on that. In essence tho, when interactions are required a short lived cookie with a reference to an "interaction session" is pushed to the user-agent and the "interaction session" is stored. The user-agent is redirected to a url resolved by a helper.

Inside interactions you load the interaction session using the `interactionDetails` helper, it accepts the request as argument as it expects the cookie there. The cookie + expected backend state also serves as CSRF and makes sure only that user-agent can resolve the interactions.

When you've satisfied the interaction based on the authorization request parameters (by storing them in the interaction session `results` property) you redirect back to the server. the `interactionResult` helper is there to do this for you, it again needs req & res to be able to manage cookies and redirect.

> can I replace stateful with stateless i.e replacing cookie/session with tokens.

No, if anything gave you the idea that was possible please bring it to my attention, it needs to be corrected.

> can I redirect to other server to make the authentication and then come back to the oauth2 server ?

Yes, given that it can still get to those cookies (domain restrictions, domain configuration see `configuration.cookies`).

You can see this in action e.g. on https://www.7pass.de, the `op` server sits on `op.7pass.de` while the interactions are an entirely different application sitting on `interactions.7pass.de`. So long as it can load and store the interaction object from the db based on the id reference in the cookie (cross checked with the uid in the request path) it's fine.

Bottom line you don't have to use the helpers, just look at what they do and emulate that in any other app.

**Comment by David-hod** [comment]
I will also put here a diagram of how I understood the server: 
![flow](https://user-images.githubusercontent.com/7628718/53962927-e33af300-40f4-11e9-83ce-334976b6b376.png)


my understanding is that each time user comes to the /interaction  we try to understand what is missing.
if authentication is missing we redirect him to **1.** if he need consent we redirect him to **2**.
if everything is fine we directly redirect him back to the client with the code/token/id_token.

am I correct ? 
If so, I don't know how the server knows if he should redirect to each of those endpoints?

**Comment by panva** [comment]
The example bundled with the current master (5.x) does not do that, it's that simple. It submits the `consent` result directly from login.

In 5.x things are tricky, you are expected to resolve all future interactions in one resume, else the OP will throw and return to client (you might've read that in the linked configuration doc). It requires you to stay in interactions until the user has consented and the example, for simplicity just submits consent results as if the user consented.

In reality you hold onto control of the user and only redirect back when he actually consents, makes sense?

In 6.x this changes, the interactions will be cyclical - you submit one prompt at a time and the OP will always ask for interaction if needed.

**Comment by panva** [comment]
> I will also put here a diagram of how I understood the server:
> ![flow](https://user-images.githubusercontent.com/7628718/53962927-e33af300-40f4-11e9-83ce-334976b6b376.png)
> 
> my understanding is that each time user comes to the /interaction we try to understand what is missing.
> if authentication is missing we redirect him to **1.** if he need consent we redirect him to **2**.
> if everything is fine we directly redirect him back to the client with the code/token/id_token.
> 
> am I correct ?
> If so, I don't know how the server knows if he should redirect to each of those endpoints?

Doesn't seem correct. The OP doesn't care about particular routes or flows, it just wants its results. How you get them in what logical order and how that's presented to the user is up to you to implement.

**Comment by panva** [comment]
How things'll work in v6.x (due sometime in April-May) you can check out its [deployment](https://guarded-cliffs-8635.herokuapp.com/.well-known/openid-configuration), you can register a client either via dynamic registration or using the client [demo site](https://tranquil-reef-95185.herokuapp.com).

https://github.com/panva/node-oidc-provider/tree/vnext is where the new provider code is.

**Comment by David-hod** [comment]
> In reality you hold onto control of the user and only redirect back when he actually consents, makes sense?

I read your message 4 times, still don't understand it.  
I also read :` app.get('/interaction/:grant', setNoCache, async (req, res, next) =>`  
if user is not login he render page that would post to  `/interaction/:grant/login` 
but I never reach `/interaction/:grant/confirm`.
does it means that I need to create my own session/token and save state for each step I have  i.e. authentication, consent ... 
when finished all steps I would call to interactionResult() with all the data that I have collected?

my guess would be that version 6 is not production ready yet. which means we cannot wait until that time. we are startup and we don't have much time :). maybe next time I will upgrade

**Comment by panva** [comment]
> does it means that I need to create my own session/token and save state for each step I have i.e. authentication, consent ...

yes

> when finished all steps I would call to interactionResult() with all the data that I have collected?

yes

> my guess would be that version 6 is not production ready yet. which means we cannot wait until that time. we are startup and we don't have much time :). maybe next time I will upgrade

`¯\_(ツ)_/¯` that's fine
